crates/wast/src/lexer.rs

```diff
@@ -266,70 +297,191 @@ impl<'a> Lexer<'a> {
     ///
     /// Returns an error if the input is malformed.
     pub fn parse(&mut self) -> Result<Option<Token<'a>>, Error> {
-        if let Some(ws) = self.ws() {
-            return Ok(Some(Token::Whitespace(ws)));
-        }
-        if let Some(comment) = self.comment()? {
-            return Ok(Some(comment));
-        }
-        if let Some(token) = self.token()? {
-            return Ok(Some(token));
-        }
-        match self.it.next() {
-            Some((i, ch)) => Err(self.error(i, LexError::Unexpected(ch))),
-            None => Ok(None),
-        }
-    }
+        let pos = self.cur();
+        // This `match` generally parses the grammar specified at
+        //
+        // https://webassembly.github.io/spec/core/text/lexical.html#text-token
+        let byte = match self.remaining.as_bytes().get(0) {
+            Some(b) => b,
+            None => return Ok(None),
+        };
 
-    fn token(&mut self) -> Result<Option<Token<'a>>, Error> {
-        // First two are easy, they're just parens
-        if let Some(pos) = self.eat_char('(') {
-            return Ok(Some(Token::LParen(&self.input[pos..pos + 1])));
-        }
-        if let Some(pos) = self.eat_char(')') {
-            return Ok(Some(Token::RParen(&self.input[pos..pos + 1])));
-        }
+        match byte {
```

Comment:
Does this block of change deserve a test? I feel the change is so big that it deserve at least a little happy path test.

Reply:
The preexisting test suite of this library was what I intended for that,  but have you found a specific bug with this PR?